### ISSUE TITLE
Which line of code reflects the pre-fetching of kv cache?

### ISSUE THREAD

**Issue by Kha-Zix-1** [snu-comparch/InfiniGen#1]
Hello！I want to know how to prefetch kv cache of the next layer to GPU while calculating the current layer.  I 'm reading `modeling_opt_ours.py`  in line 795

``` python
if (idx > 0) and (idx < (len(self.layers)-1)):
                    cur_device = self.layers[idx].self_attn.current_hidden_states.device
                    cur_dtype = self.layers[idx].self_attn.current_hidden_states.dtype
                    cur_bsz, cur_tgt_len, _ = self.layers[idx].self_attn.current_hidden_states.shape
                    self.layers[idx + 1].self_attn.previous_hidden_states = torch.cat((self.layers[idx].self_attn.current_hidden_states, 
                                                                                       torch.ones(cur_bsz, cur_tgt_len, 1).to(cur_device).to(cur_dtype)), dim = -1)
```
You load the next layer's hidden_states to GPU for the next layer to calculate the speculative attention. However, I don't think this way overlaps the prefetching process and the calculating process. On the contrary, it just moves the kv cache loading process of the next layer to the current layer. To me, a prefetch process should create a new thread or a new stream to make the code execute asynchronously. So could you please help me understand how to **prefetch** the kv cache? Many thanks!

**Comment by leewonbeom** [comment]
Hi @Kha-Zix-1, 
Thank you for your interest in our work.

We have two repositories: one for `accuracy` and another for `speedup` evaluation. The `accuracy` directory (including `modeling_opt_ours.py`) focuses on implementing the functionality of InfiniGen (i.e., the correctness of the computation) to enable efficient accuracy evaluation.

The `speedup` directory contains the codes for prefetching using a new stream, as you mentioned. You can find it in the line 461 of `speedup/flexgen/infinigen/flex_opt.py`. I think reading that code might be helpful to understand how InfiniGen performs prefetching.

Hope this helps!

**Comment by Kha-Zix-1** [comment]
Many thanks!
